nixos/modules/services/hardware/thermald.nix

```diff
@@ -4,27 +4,392 @@ with lib;
+        '';
+        default = {};
+        type = submodule {
+          options = {
+            kp = mkOption {
+              type = string;
+              description = "kp";
+            };
+
+            kd = mkOption {
+              type = string;
+              description = "kd";
+            };
+
+            ki = mkOption {
+              type = string;
+              description = "ki";
+            };
+          };
+        };
+      };
+
+      WritePrefix = mkOption {
+        type = nullOr string;
+        default = null;
+        description = ''
+          Write some prefix attached to state value, like below the prefix is "level ". It will
+          preserve spaces as entered when writing to sysfs
+        '';
+      };
+    };
+  };
+
+  makeThermalConfigurationXml = configFile:
+    "<?xml version=\"1.0\"?>\n" +
+    xmlBlock "ThermalConfiguration" (map makePlatformXml configFile.platforms);
+
+  makePlatformXml = pCfg: xmlBlock "Platform" [
+    (xmlElement pCfg "Name")
+    (xmlElement pCfg "UUID")
+    (xmlElement pCfg "ProductName")
+    (xmlElement pCfg "Preference")
+    (xmlBlock "ThermalSensors" (map makeThermalSensorXml pCfg.ThermalSensors))
+    (xmlBlock "ThermalZones" (map makeThermalZoneXml pCfg.ThermalZones))
+    (xmlBlock "CoolingDevices" (map makeCoolingDeviceXml pCfg.CoolingDevices))
+  ];
+
+  makeThermalSensorXml = tsCfg: xmlBlock "ThermalSensor" [
+    (xmlElement tzCfg "Type")
+    (xmlElement tzCfg "Path")
+    (mapXmlElement boolToInt tzCfg "AsyncCapable")
+  ];
+
+  makeThermalZoneXml = tzCfg: xmlBlock "ThermalZone" [
+    (xmlElement tzCfg "Type")
+    (xmlElement tzCfg "Path")
+    (mapXmlElement boolToInt tzCfg "AsyncCapable")
+    (xmlBlock "TripPoints" (map makeTripPointXml tzCfg.TripPoints))
+  ];
+
+  makeTripPointXml = tpCfg: xmlBlock "TripPoint" ([
+    (xmlElement tpCfg "SensorType")
+    (mapXmlElement toString tpCfg "Temperature")
+    (xmlElement tpCfg "Type")
+    (xmlElement tpCfg "ControlType")
+  ] ++ (map makeTripPointCoolingDeviceXml tpCfg.CoolingDevices));
+
+  makeTripPointCoolingDeviceXml = tpcdCfg: xmlBlock "CoolingDevice" [
+    (mapXmlElement toString tpcdCfg "Index")
+    (xmlElement tpcdCfg "Type")
+    (mapXmlElement toString tpcdCfg "Influence")
+    (mapXmlElement toString tpcdCfg "SamplingPeriod")
+    (mapXmlElement toString tpcdCfg "TargetState")
+  ];
+
+  makeCoolingDeviceXml = cdCfg: xmlBlock "CoolingDevice" [
+    (xmlElement cdCfg "Type")
+    (mapXmlElement toString cdCfg "MinState")
+    (mapXmlElement toString cdCfg "IncDecStep")
+    (mapXmlElement toString cdCfg "ReadBack")
+    (mapXmlElement toString cdCfg "MaxState")
+    (mapXmlElement toString cdCfg "DebouncePeriod")
+    (xmlBlock "PidControl" [
+      (xmlElement cdCfg.PidControl "kp")
+      (xmlElement cdCfg.PidControl "kd")
+      (xmlElement cdCfg.PidControl "ki")
+    ])
+    (xmlElement cdCfg "WritePrefix")
+  ];
```

Comment:
I think a generic xml generator would be better than this. Something like

```nix
{
  config = {
    ThermalConfiguration.Platform = map (platform: {
      Name = platform.name;
      UUID = platform.UUID;
      # ...
      TermalSensors.ThermalSensor = map (sensor: {
        Name = sensor.name;
        # ...
      }) platform.thermalSensors;
      # ...
    }) configFile.platforms;
  };
  
  toXml = val:
    # Generate xml value by using the type of `val`:
    # toString for simple values
    # for attrsets use name/value as xml attr and value (recursively). if value is a list, generate multiple with the same name
  ;

  configXml = "<?xml version..." + concatStrings (toXml cfg.config);
}
```

You might want to make `toXML` return a list of strings, representing lines of the config file, then you can easily indent them when you recurse by mapping over them.

I did that for my ZNC PR (its syntax is not xml unfortunately): https://github.com/NixOS/nixpkgs/pull/41467/files#diff-f93a99c1b50dc7cd526921714cb77de1R22

Reply:
I had a go at using a generic representation but I've hit a snag that I'm not sure how to resolve.

Currently I've written `toXML` like so:

```
  toXml =
    let listToXml = list: concatLines (map toXml' list);
        attrToXml = key: value:
          if skipAttribute value
          then ""
          else "<${key}>" + (indentIfNeeded value (toXml' value)) + "</${key}>";
        setToXml = set: (concatLines (map (k: attrToXml k (set.${k})) (builtins.attrNames set)));

        # We don't want to include empty lines in the output as it bloats the XML for no reason.
        concatLines = lines: builtins.concatStringsSep "\n" (lib.remove "" lines);

        # We want to completel remove some attributes based on their value as they provide no value
        # to the output and cause bloating if left in.
        skipAttribute = value: value == null || value == [];

        # We don't want to indent strings since they're usually short one-liners
        skipIndent = value: (builtins.typeOf value == "string");
        indentIfNeeded = value: xml:
          if skipIndent value then xml else "\n" + (indent xml) + "\n";

        toXml' = value:
                if builtins.typeOf value == "list" then listToXml value
           else if builtins.typeOf value == "set"  then setToXml value
           else builtins.toString value;
    in toXml'
```

This works fine for the mots of the configuration but there's a particular configuration detail that I don't know how to represent using sets.

When defining multiple `CoolingDevice`s in a trip point the XML expects the following format:

```
<TripPoint>
  ...
  <CoolingDevice>
    ...
  </CoolingDevice>
  <CoolingDevice>
    ...
  </CoolingDevice>
</TripPoint>
```

This is different to every other list in the config file in that it _doesn't_ have a parent `<CoolingDevices>` tag. As a result I would need some way to indicate in a set that multiple keys should be merged inline. I'd imagine it would look something like this:

```
tripPoint: {
  CoolingDevice = ...
  CoolingDevice = ...
}
```

But as far as I understand it, sets cannot have duplicate entries.

In the current solution I solve it by introducing a `CoolingDevices` attribute and then interpreting that as a special case:

```
  makeTripPointXml = tpCfg: xmlBlock "TripPoint" ([
    (xmlElement tpCfg "SensorType")
    (mapXmlElement toString tpCfg "Temperature")
    (xmlElement tpCfg "Type")
    (xmlElement tpCfg "ControlType")
  ] ++ (map makeTripPointCoolingDeviceXml tpCfg.CoolingDevices));
```

However I don't believe I can do this with a generic builder.

Please let me know if there's a way around this, otherwise I think it might make sense to stick with the `xmlElement` and `xmlBlock` based approach.